File: kotlin/src/main/kotlin/adventofcode2022/Day02.kt
package adventofcode2022

import adventofcode2022.Hand.Companion.getWinningHand
import java.io.File

fun day02() {
    val file = File("./input/day02.txt")
    val wrongHands = file.readLines().stream()
            .map {
                val split = it.split(" ")
                Pair(Hand.from(split[0]), Hand.from(split[1]))
            }.toList()
    var wrongPoints = 0
    for (hand in wrongHands) {
        wrongPoints += play(hand)
    }
    println("Total points from wrong strategy: $wrongPoints")

    val correctHands = file.readLines().stream().map {
        val split = it.split(" ")
        Pair(Hand.from(split[0]), convertToCorrectHand(split))
    }.toList()
    var correctPoints = 0
    for (hand in correctHands) {
        correctPoints += play(hand)
    }
    println("Total points from correct strategy: $correctPoints")
}

fun convertToCorrectHand(strategy: List<String>): Hand {
    val goForStrategy = strategy[1]
    val opponentHand = Hand.from(strategy[0])
    if (goForStrategy == "Z") {
        return getWinningHand(opponentHand)
    }
    if (goForStrategy == "Y") {
        return opponentHand
    }
    return getWinningHand(getWinningHand(opponentHand))
}

enum class Hand {
    ROCK, PAPER, SCISSOR, NONE;
    companion object {
        fun from(code: String): Hand {
            return when(code) {
                "X", "A" -> ROCK
                "Y", "B" -> PAPER
                "Z", "C" -> SCISSOR
                else -> NONE
            }
        }

        fun getWinningHand(hand: Hand): Hand {
            return when (hand) {
                ROCK -> PAPER
                PAPER -> SCISSOR
                SCISSOR -> ROCK
                else -> NONE
            }
        }
    }
}

fun play(hands: Pair<Hand, Hand>): Int {
    if (isWinningHand(hands)) {
        return getPointsForResult(left = false, right = true) + getPointForHand(hands.second)
    }
    if (isDrawingHand(hands)) {
        return getPointsForResult(left = true, right = true) + getPointForHand(hands.second)
    }
    return 0 + getPointForHand(hands.second)
}

private fun isWinningHand(hands: Pair<Hand, Hand>) = listOf(
        Pair(Hand.SCISSOR, Hand.ROCK),
        Pair(Hand.ROCK, Hand.PAPER),
        Pair(Hand.PAPER, Hand.SCISSOR)
).contains(hands)

fun isDrawingHand(hands: Pair<Hand, Hand>) = listOf(
        Pair(Hand.ROCK, Hand.ROCK),
        Pair(Hand.PAPER, Hand.PAPER),
        Pair(Hand.SCISSOR, Hand.SCISSOR)
).contains(hands)

fun getPointsForResult(left: Boolean, right: Boolean): Int {
    if (right && !left) {
        return 6
    }
    if (right == left) {
        return 3
    }
    return 0
}

fun getPointForHand(hand: Hand): Int {
    return when(hand) {
        Hand.ROCK -> 1
        Hand.PAPER -> 2
        Hand.SCISSOR -> 3
        else -> 0
    }
}

fun main() {
    val startTime = System.nanoTime()
    day02()
    val endTime = System.nanoTime()
    println("Elapsed time: ${(endTime - startTime).toDouble() / 1000000000} sec")
}

File: python/Day02.py

from enum import Enum

from utils.Stopwatch import runWithStopwatch


class Hand(Enum):
    ROCK = 1
    PAPER = 2
    SCISSOR = 3

    @staticmethod
    def of(code: str):
        match code:
            case "A" | "X":
                return Hand.ROCK
            case "B" | "Y":
                return Hand.PAPER
            case "C" | "Z":
                return Hand.SCISSOR
            case other:
                return 0

    def getWinningHand(self):
        match self:
            case Hand.ROCK:
                return Hand.PAPER
            case Hand.PAPER:
                return Hand.SCISSOR
            case Hand.SCISSOR:
                return Hand.ROCK
            case other:
                return 0


def play(opponent: Hand, hand: Hand):
    if opponent.getWinningHand() == hand:
        return 6 + hand.value
    elif opponent == hand:
        return 3 + hand.value
    return hand.value


def convertToHand(opponent_hand: Hand, strategy: str) -> Hand:
    match strategy:
        case "X":
            return opponent_hand.getWinningHand().getWinningHand()
        case "Y":
            return opponent_hand
        case "Z":
            return opponent_hand.getWinningHand()


def day02():
    file = open("../input/day02.txt", "r")
    lines = file.readlines()
    wrong_points = 0
    correct_points = 0
    for line in lines:
        split = line.strip().split(" ")
        opponent_hand = Hand.of(split[0])
        wrong_points += play(opponent_hand, Hand.of(split[1]))
        correct_points += play(opponent_hand, convertToHand(opponent_hand, split[1]))
    print("Total wrong points: " + str(wrong_points))
    print("Total correct points: " + str(correct_points))


if __name__ == '__main__':
    runWithStopwatch(day02)
